drivers/console/ipm_console_receiver.c

```diff
@@ -66,24 +66,12 @@ static void ipm_console_thread(void *arg1, void *arg2, void *arg3)
 			++pos;
 		}
 
-		/* ISR may have disabled the channel due to full buffer at
-		 * some point. If that happened and there is now room,
-		 * re-enable it.
-		 *
-		 * Lock interrupts to avoid pathological scenario where
-		 * the buffer fills up in between enabling the channel and
-		 * clearing the channel_disabled flag.
-		 */
-		if (driver_data->channel_disabled &&
-		    ring_buf_space_get(&driver_data->rb)) {
-			key = irq_lock();
-			ipm_set_enabled(driver_data->ipm_device, 1);
-			driver_data->channel_disabled = 0;
-			irq_unlock(key);
-		}
 	}
 }
 
+#ifdef CONFIG_SMP
+__attribute__((optimize("O0")))
```

Comment:
for  "Broadly though, I don't understand why this fixes what it says it fixes. How exactly was this broken in SMP? The commit message should have an explanation."

And prior this PR, with SMP on, there is the possibility that offload_irq and rx_thread run at different cpu core at the same time, it will access ipm_set_enable and channel_disable at the same time and causing state(enabled and channel_disable) wrong. And this PR just use two threads(main thread and rx thread) and sync up each other with sem and ring buffer's space. It will never access same thing at the same time.